source/val/ValidationState.cpp

```diff
@@ -357,4 +365,91 @@ spv_result_t ValidationState_t::RegisterFunctionEnd() {
   return SPV_SUCCESS;
 }
 
+ValidationState_t::UseDefTracker::UseDefTracker(ValidationState_t* state)
+    : module_(state) {}
+
+void ValidationState_t::UseDefTracker::AddDef(const spv_id_info_t& def) {
+  all_defs_[def.id] = def;
+  if (module_->in_block()) {
+    auto block = module_->get_current_function()->get_current_block();
+    block_defs_[block->get_id()].push_back(def.id);
+  } else {
+    module_defs_.insert(def.id);
+  }
+}
+
+void ValidationState_t::UseDefTracker::AddUse(uint32_t id, bool is_phi) {
+  all_uses_.insert(id);
+  if (module_->in_block()) {
+    auto block = module_->get_current_function()->get_current_block();
+    block_uses_[block->get_id()].insert({id, is_phi});
+  }
+}
+
+std::pair<bool, spv_id_info_t> ValidationState_t::UseDefTracker::FindDef(
+    uint32_t id) const {
+  if (all_defs_.count(id) == 0) {
+    return std::make_pair(false, spv_id_info_t{});
+  } else {
+    /// We are in a const function, so we cannot use defs.operator[]().
+    /// Luckily we know the key exists, so defs_.at() won't throw an
+    /// exception.
+    return std::make_pair(true, all_defs_.at(id));
+  }
+}
+
+std::unordered_set<uint32_t>
+ValidationState_t::UseDefTracker::FindUsesWithoutDefs() const {
+  auto diff = all_uses_;
+  for (const auto d : all_defs_) diff.erase(d.first);
+  return diff;
+}
+
+void ValidationState_t::UseDefTracker::UpdateDefs(Function* func) {
+  auto& ordered_blocks = func->get_blocks();
+
+  // Because dominators must appear before the blocks they dominate in the
+  // binary , we only need to visit the immidiate dominators to propagate the
+  // ids throughout the CFG
+  for (auto block : ordered_blocks) {
+    auto& block_defs = block_defs_[block->get_id()];
+    auto idom = block->GetImmediateDominator();
+    if (block->is_reachable() && block != idom) {
+      assert(idom);
+      auto& idom_defs = block_defs_[idom->get_id()];
+      block_defs.reserve(block_defs.size() + idom_defs.size());
+      copy(begin(idom_defs), end(idom_defs), back_inserter(block_defs));
+    } else {
+      // This is the first block or it is unreachable in the CFG
+      block_defs.reserve(block_defs.size() + module_defs_.size());
+      copy(begin(module_defs_), end(module_defs_), back_inserter(block_defs));
```

Comment:
This is not quite right.
OpFunctionParameter values are defined outside any block, so they incorrectly end up in the module_defs_ bucket.
So this code will allow an OpFunctionParameter from one function definition to be used in the body of a later function definition.

Example:

```
lOpCapability Shader
OpMemoryModel Logical Simple

%void = OpTypeVoid
%i32 = OpTypeInt 32 1
%voidfn = OpTypeFunction %void %i32

%f0 = OpFunction %void None %voidfn
%a0 = OpFunctionParameter %i32
%entry0 = OpLabel
OpReturn
OpFunctionEnd

%f1 = OpFunction %void None %voidfn
%a1 = OpFunctionParameter %i32
%entry1 = OpLabel
%oops = OpCopyObject %i32 %a0  ; this should not be allowed
OpReturn
OpFunctionEnd
```